test-services/test-probe/src/context.rs

```diff
@@ -0,0 +1,72 @@
+//! Define routes for exposing the contents of the application's
+//! rendering context.
+
+use crate::{config,
+            error::Result};
+use actix_web::{web,
+                HttpRequest,
+                HttpResponse};
+use serde_json;
+use std::{fs,
+          path::{Components,
+                 Path,
+                 PathBuf}};
+
+/// Show render context rooted at `path`
+pub fn show(config: web::Data<config::Config>,
+            req: HttpRequest)
+            -> actix_web::Result<HttpResponse> {
+    let path = req.match_info()
+        .get("path")
+        .map(PathBuf::from)
+        .unwrap_or_default();
+    let json =
+        extract_data(&config.render_context_file, path).map_err(actix_web::error::ErrorBadRequest)?;
+
+    if let Some(json) = json {
+        Ok(HttpResponse::Ok().json(json))
+    } else {
+        Ok(HttpResponse::NotFound().finish())
+    }
+}
+
+////////////////////////////////////////////////////////////////////////
+
+fn extract_data<P>(render_context_file: P, path: PathBuf) -> Result<Option<serde_json::Value>>
+    where P: AsRef<Path>
+{
+    let data = read_json(render_context_file)?;
+    lookup(data, path.components())
+}
+
+/// Read a file into a JSON value
+fn read_json<P>(path: P) -> Result<serde_json::Value>
+    where P: AsRef<Path>
+{
+    let contents = fs::read_to_string(path)?;
+    serde_json::from_str(&contents).map_err(|e| e.into())
+}
+
+/// Recursively descend through a JSON structure, based on the path
+/// provided in `keys`
+fn lookup(data: serde_json::Value, keys: Components) -> Result<Option<serde_json::Value>> {
```

Comment:
This is clever! Could we use [`pointer`](https://docs.serde.rs/serde_json/enum.Value.html#method.pointer)?

Reply:
Whoa, nice... I didn't realize that was a thing. A pointer definitely seems more in the spirit of what I was going for.

I'll check that out!

Reply:
_Much_ better: https://github.com/habitat-sh/habitat/pull/7283/commits/29c091c026512a46fdce46a1dde75d51895f9f11